File: src/parsers/course-parser.ts
import * as cheerio from 'cheerio';
import type { Course } from '../types/course';
import { noopLogger, type Logger } from '../utils/logger';

export function parseCoursesFromHome(html: string, debug: Logger = noopLogger): Course[] {
  const log = debug;
  const $ = cheerio.load(html);
  const seen = new Set<string>();
  const courses: Course[] = [];

  log.debug(`parseCoursesFromHome: HTML length=${html.length}`);

  $('a[href*="Course.aspx"]').each((_, el) => {
    const href = $(el).attr('href') || '';
    const text = $(el).text().trim();

    if (!text && !href) return;

    const codeFromHref = extractCodeFromHref(href);
    const codeFromText = extractCodeFromText(text);
    const code = codeFromHref || codeFromText;
    if (!code || seen.has(code)) return;

    seen.add(code);
    courses.push({
      code,
      title: extractTitle(text, code),
    });
    log.debug(`  Course from link: ${code} - ${extractTitle(text, code)}`);
  });

  if (courses.length > 0) {
    log.info(`parseCoursesFromHome: ${courses.length} courses from links`);
    return courses;
  }

  $('[id^="MainContent_gvCourseList_ibtnAssignments_"]').each((_: number, el) => {
    const card = $(el).closest('.m-portlet');
    const h3 = card.find('h3').first();
    const h3Text = h3.text().trim();
    const code = extractCodeFromText(h3Text);
    if (code && !seen.has(code)) {
      seen.add(code);
      courses.push({
        code,
        title: extractTitle(h3Text, code),
      });
      log.debug(`  Course from portlet: ${code} - ${extractTitle(h3Text, code)}`);
    }
  });

  log.info(`parseCoursesFromHome: ${courses.length} courses from portlets`);
  return courses;
}

function extractCodeFromHref(href: string): string | null {
  const match = href.match(/[?&]code=([^&]+)/i);
  return match ? match[1].toUpperCase() : null;
}

function extractCodeFromText(text: string): string | null {
  const match = text.match(/^([A-Z]{2,4}\d{3}[A-Z]?)\b/i);
  return match ? match[1].toUpperCase() : null;
}

function extractTitle(fullText: string, code: string): string {
  const escaped = code.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const title = fullText.replace(new RegExp(`^${escaped}\\s*-?\\s*`, 'i'), '').trim();
  const firstLine = title.split(/\n/)[0].trim();
  return firstLine || code;
}

File: vulms_dart/lib/src/parsers/course_parser.dart
import 'package:html/parser.dart' as html_parser;

import '../models/course.dart';
import '../utils/logger.dart';
import '../utils/dom_helper.dart';

/// Parses course information from VULMS HTML.
class CourseParser {
  CourseParser._();

  /// Parse enrolled courses from Home.aspx HTML.
  static List<Course> parseFromHome(String html, {VulmsLogger? logger}) {
    final doc = html_parser.parse(html);
    final seen = <String>{};
    final courses = <Course>[];

    logger?.debug('parseCoursesFromHome: HTML length=${html.length}');

    // Try course links first
    final links = doc.querySelectorAll('a[href*="Course.aspx"]');
    for (final el in links) {
      final href = el.attributes['href'] ?? '';
      final text = el.text.trim();
      if (text.isEmpty && href.isEmpty) continue;

      final code = _extractCodeFromHref(href) ?? _extractCodeFromText(text);
      if (code == null || seen.contains(code)) continue;

      seen.add(code);
      courses.add(Course(code: code, title: _extractTitle(text, code)));
      logger?.debug('  Course from link: $code - ${_extractTitle(text, code)}');
    }

    if (courses.isNotEmpty) {
      logger?.info('parseCoursesFromHome: ${courses.length} courses from links');
      return courses;
    }

    // Fallback: course cards from dashboard
    final assignmentButtons =
        doc.querySelectorAll('[id^="MainContent_gvCourseList_ibtnAssignments_"]');
    for (final el in assignmentButtons) {
      final card = closest(el, '.m-portlet');
      if (card == null) continue;
      final h3 = card.querySelector('h3');
      if (h3 == null) continue;
      final h3Text = h3.text.trim();
      final code = _extractCodeFromText(h3Text);
      if (code != null && !seen.contains(code)) {
        seen.add(code);
        courses.add(Course(code: code, title: _extractTitle(h3Text, code)));
        logger?.debug(
            '  Course from portlet: $code - ${_extractTitle(h3Text, code)}');
      }
    }

    logger?.info(
        'parseCoursesFromHome: ${courses.length} courses from portlets');
    return courses;
  }

  static String? _extractCodeFromHref(String href) {
    final match = RegExp(r'[?&]code=([^&]+)', caseSensitive: false)
        .firstMatch(href);
    return match?.group(1)?.toUpperCase();
  }

  static String? _extractCodeFromText(String text) {
    final match = RegExp(r'^([A-Z]{2,4}\d{3}[A-Z]?)\b').firstMatch(text);
    return match?.group(1)?.toUpperCase();
  }

  static String _extractTitle(String fullText, String code) {
    final escaped = code.replaceAll(RegExp(r'[.*+?^${}()|[\]\\]'), r'\\$&');
    final title = fullText
        .replaceFirst(RegExp('^$escaped\\s*-?\\s*', caseSensitive: false), '')
        .trim();
    final firstLine = title.split('\n').first.trim();
    return firstLine.isEmpty ? code : firstLine;
  }
}
